apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ArchiveApiTypeContainer.java

```diff
@@ -149,31 +115,46 @@ public String toString() {
 	String fLocation;
 
 	/**
-	 * Cache of package names to class file paths in that package, or
-	 * <code>null</code> if not yet initialized.
+	 * Cache of package names to a map of class names to class files paths in that
+	 * package, or <code>null</code> if not yet initialized.
 	 */
-	private Map<String, Set<String>> fPackages;
+	private Map<String, Map<String, String>> fPackages;
 
 	/**
 	 * Cache of package names in this archive.
 	 */
 	private String[] fPackageNames;
 
 	/**
-	 * Constructs an {@link IApiTypeContainer} container for the given jar or
-	 * zip file at the specified location.
+	 * Constructs an {@link IApiTypeContainer} container for the given jar or zip
+	 * file at the specified location.
 	 *
 	 * @param parent the parent {@link IApiElement} or <code>null</code> if none
-	 * @param path location of the file in the local file system
+	 * @param path   location of the file in the local file system
 	 */
 	public ArchiveApiTypeContainer(IApiElement parent, String path) {
 		super(parent, IApiElement.API_TYPE_CONTAINER, path);
 		this.fLocation = path;
-		if (path.endsWith("jrt-fs.jar")) { //$NON-NLS-1$
-			IPath newPath = new Path(path);
-			newPath = newPath.removeLastSegments(2).addTrailingSeparator();
-			newPath = newPath.append("jmods").append("java.base.jmod"); //$NON-NLS-1$ //$NON-NLS-2$
-			this.fLocation = newPath.toOSString();
+	}
+
+	/**
+	 * Converts the location to a path in the applicable file system.
+	 *
+	 * @return the path corresponding to the location.
+	 * @throws IOException
+	 */
+	@SuppressWarnings("nls")
+	private Path getLocation() throws IOException {
+		if (fLocation.endsWith("jrt-fs.jar")) {
+			Path jrePath = Path.of(fLocation).getParent().getParent();
+			try (URLClassLoader loader = new URLClassLoader(new URL[] { jrePath.toUri().toURL() })) {
```

Comment:
No idea what is the life cycle of this container is, and how often getLocation() is called, but isn't this change leaking file system/classloader every time (classloader will stay there as long as loaded classes are not GC'd)? Shouldn't the result be cached?

I assume you use new classloader to avoid "FS already created" error? But it would be OK to keep the FS open and re-use existing one if there is one already.

Reply:
I see other usage patterns similar to this, but it's better safe than sorry and create only one instance per JRE location.  That will perform better as well in any case.  I'll update the PR accordingly.  I need to fix versions anyway...  Thanks for the early feedback.  😄

Reply:
Note that the class loader itself is closed.  I actually wonder if I shouldn't be passing the system class loaded... 

Let's see how the tests go and I'll review that consideration when I'm back.

Reply:
The latest force push doesn't use a class loader at all, i.e., like how org.eclipse.jdt.internal.compiler.util.JrtFileSystem.JrtFileSystem(Jdk, String) does it.